Custom Authentication Mechanisms for WebDAV and APIs

joneug opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->
The authentication mechanisms for securing WebDAV and APIs were extended to allow 3rd-party apps to provide additional authentication backends. In the WebDAV interface the dispatching of the `OCA\DAV\Connector\Sabre::authInit` event was added. For securing the API the interface `IAuthModule` was added. Apps can register their implementation of this interface in the `info.xml` file.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
#10400

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
This, for example, allows the implementation of Bearer Authentication with OAuth 2.0 (see [OAuth 2.0 app](https://github.com/owncloud/oauth2)).

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
The OAuth2 app's PHPUnit tests have been run on [Travis CI](https://travis-ci.org/pssl16/owncloud-app_oauth2) with the following configurations:
- PHP versions: 5.6, 7
- Core branches: master
- Databases: PostgreSQL, MySQL, SQLite

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

joneug committed:
Add if-statement for switching auth backends

joneug committed:
Implement determineUsername function

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/core?pullRequest=26742) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/core?pullRequest=26742) before we can accept your contribution.<br/><hr/>**Jonathan Neugebauer** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

DeepDiver1975 commented:
@PhilippSchaffrath @Peter-Prochaska

joneug committed:
Update PHPDoc

joneug committed:
Update code for changes in OAuth 2.0 app

joneug committed:
Replace own AbstractBearer implementation with that of sabre/dav

guruz commented:
Linking https://github.com/owncloud/core/issues/24995

joneug committed:
Add OAuth 2.0 App as submodule

joneug committed:
Update comment in webdav.php

joneug committed:
Improve if-statement for the auth backend

joneug committed:
Update code for changes in OAuth 2.0 app

joneug committed:
Add OAuth2Test

joneug commented:
@PhilippSchaffrath @DeepDiver1975 @butonic Please review the changes. Looking forward to your ideas.

joneug committed:
Update OAuth2Test

joneug commented:
I think the problem with the build failing in Jenkins is that the OAuth 2.0 app was added as a [Submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules) which requires additional commands for cloning. Would you like to use submodules or should I include the OAuth 2.0 app's files directly?

joneug committed:
Merge branch 'master' of https://github.com/owncloud/core into dav-oauth

joneug committed:
Add OAuth 2.0 login mechanism to Session.php for use in api.php

joneug committed:
Update loginWithOAuth2 function
User Hooks are currently not triggered because of the missing password

joneug committed:
Update OAuth 2.0 App

joneug committed:
Improve tearDown function in OAuth2Test

joneug committed:
Refactor OAuth2Test

joneug committed:
Refactor OAuth2Test

DeepDiver1975 commented:
> I think the problem with the build failing in Jenkins is that the OAuth 2.0 app was added as a Submodule which requires additional commands for cloning. Would you like to use submodules or should I include the OAuth 2.0 app's files directly?

Linking the app via submodule is not what we really want.

What about moving your app into the owncloud organisation and also move the sabre dav related code from this PR into the app as well?

DeepDiver1975 commented on apps/dav/appinfo/v1/webdav.php:
```diff
@@ -40,13 +40,20 @@
-	\OC::$server->getTwoFactorAuthManager(),
-	'principals/'
-);
+if (strpos(explode(' ', \OC::$server->getRequest()->getHeader('Authorization'))[0], 'Bearer') !== false) {
```
We need some registration/factory mechanism to instantiate the correct auth backend - I'm happy to help out here.

joneug replied:
Looking forward to your ideas.

DeepDiver1975 commented on apps/dav/appinfo/v1/webdav.php:
```diff
@@ -40,13 +40,20 @@
-);
+if (strpos(explode(' ', \OC::$server->getRequest()->getHeader('Authorization'))[0], 'Bearer') !== false) {
+	// OAuth 2.0
+	$authBackend = new \OCA\DAV\Connector\Sabre\OAuth2();
```
You should really inject all dependencies in the ctor

joneug committed:
Remove submodule oauth2

joneug committed:
Update .gitignore

joneug commented:
> Linking the app via submodule is not what we really want.

I removed the submodule.

> What about moving your app into the owncloud organisation and also move the sabre dav related code from this PR into the app as well?

I moved the code related to sabre/dav into the app. What are your thoughts on the `dav` app being dependent on the `oauth2` app when trying to support both authentication backends? Are you planning to include the `oauth2` app in the core?

joneug committed:
Merge branch 'dav-oauth' into api-oauth

DeepDiver1975 commented:
> Are you planning to include the oauth2 app in the core?

that could be one options ....



> What are your thoughts on the dav app being dependent on the oauth2 app when trying to support both authentication backends?

We have to avoid such a dependency - what you can try is to use the event handler
https://github.com/owncloud/core/blob/master/apps/dav/lib/Server.php#L85-L87

To get an idea on how to use this: https://github.com/owncloud/core/blob/master/apps/federation/lib/AppInfo/Application.php#L124-L135


What is not yet covered in this approach is to allow bearer tokens to aallow access to the OCS api.
We can address this in a second stage

joneug commented:
> We have to avoid such a dependency - what you can try is to use the event handler

I will look into that. Thanks!

> What about moving your app into the owncloud organisation

I tried to [transfer the repository](https://help.github.com/articles/transferring-a-repository-owned-by-your-organization/) but I don't have the rights to add a repository in ownCloud. Could you please provide a repository?

joneug committed:
Remove submodule oauth2

joneug committed:
Delete OAuth2Test.php

DeepDiver1975 commented:
> I tried to transfer the repository but I don't have the rights to add a repository in ownCloud. Could you please provide a repository?

Please transfer the repo to my - I can move it to the owncloud organization

joneug commented:
> What is not yet covered in this approach is to allow bearer tokens to aallow access to the OCS api.
We can address this in a second stage

I added some commits regarding this.

joneug commented:
> Please transfer the repo to my - I can move it to the owncloud organization

Done.

joneug commented on lib/private/User/Session.php:
```diff
@@ -516,6 +520,59 @@ private function loginWithToken($token) {
+		$userId = $accessToken->getUserId();
+
+		// TODO: Where to get the password for the preLogin hook?
+		//$this->manager->emit('\OC\User', 'preLogin', [$userId, '']);
```
Where to get the password for the preLogin hook?

joneug commented on lib/private/User/Session.php:
```diff
@@ -516,6 +520,59 @@ private function loginWithToken($token) {
+		$this->setLoginName($user->getDisplayName());
+
+		// TODO: Where to get the password for the postLogin hook?
+		//$this->manager->emit('\OC\User', 'postLogin', [$userId, '']);
```
Where to get the password for the postLogin hook?

DeepDiver1975 replied:
On creation of the oauth token you could store the password - this is what is done in the token based authentication as implemented in core since 9.1

Not really nice to store the password - but we need the password in many areas ....

joneug replied:
But in the `oauth2` app, where the Access Tokens are being generated, we are not handling passwords at all. This is all done by the API's security mechanism...

joneug replied:
Besides that an AuthModule only provides the User. So inside the `loginUser` function, where the hooks should be triggered, we still don't have the password.

What do you think?

joneug replied:
@DeepDiver1975 If storing the password is the only way, we would need to pass the password to the app after the user logged in for the client's Authorization Request (in `LoginController`'s `tryLogin` function). But besides introducing another dependency, to me it does not seem to be a good idea to pass passwords to apps.

Any ideas?

DeepDiver1975 commented:
> Done.

@joneug I did not receive anything yet - mind checking again? THX

phisch commented:
@joneug we are planning on relying on the oauth implementation, so we should include everything in core instead of an app. (for example the clients will only use oauth in the future)

Have you checked out existing oauth server libraries?
One that looks quite well developed is this one: https://github.com/thephpleague/oauth2-server

Although we will not be able to use it to it's full potential, since we can't support middleware right now. But it should provide enough to get a solid implementation right now, and clean up easily once we decide to switch to middleware compatible psr-7 request/responses.

If you need any help, or have any questions, just ping me, i will do my best to help you getting this done!

joneug commented:
> @joneug I did not receive anything yet - mind checking again? THX

This is odd. I aborted the existing request and opened a new one. Hopefully now it works...

DeepDiver1975 commented:
> This is odd. I aborted the existing request and opened a new one. Hopefully now it works...

here we go - https://github.com/owncloud/oauth2

DeepDiver1975 commented:
> we are planning on relying on the oauth implementation, so we should include everything in core instead of an app. (for example the clients will only use oauth in the future)

We need to change pieces within core to make it work this is quite true - but generally speaking building this as an app is quite valid - also an app in its own repo.

phisch commented:
Even if it is an app that is always enabled, i just don't see the benefit from putting it into an app. Getting this to work with all different request entry points could be quite hook-heavy if it is implemented in an app. Doing this in core directly should make the implementation simpler.

DeepDiver1975 commented:
> Doing this in core directly should make the implementation simpler.

for sure not - e.g. the whole dav implementation is living in an app of it's own - https://github.com/owncloud/core/tree/master/apps/dav

joneug commented:
@DeepDiver1975 Could you please add Travis and codecov integration for the new repo?

joneug commented:
> We have to avoid such a dependency - what you can try is to use the event handler

I tried to add the event listener in the constructor of the app's `Application.php` like this:

```php
$dispatcher = $this->getContainer()->getServer()->getEventDispatcher();
		$dispatcher->addListener('OCA\DAV\Connector\Sabre::authInit', function($event) use($container) {
			if ($event instanceof SabrePluginEvent) {
				$authPlugin = $event->getServer()->getPlugin('auth');
				if ($authPlugin instanceof Plugin) {
					$authPlugin->addBackend(new OAuth2());
				}
			}
		});
```

Unfortunately I did not have luck authenticating after that. I am assuming that `/remote.php/dav/` is the right endpoint since only `appinfo/v2/remote.php` is using `\OCA\DAV\Server`.

DeepDiver1975 commented:
> Unfortunately I did not have luck authenticating after that. I am assuming that /remote.php/dav/ is the right endpoint since only appinfo/v2/remote.php is using \OCA\DAV\Server.

yes - this is only for remote.php/dav - a similar mechanism has to be added to remote.php/webdav.
Any also for the OCS API controllers - maybe we need a different approach here.

Let me think about this ....

DeepDiver1975 commented on lib/private/User/Session.php:
```diff
@@ -516,6 +520,59 @@ private function loginWithToken($token) {
+	 * @return boolean
+	 * @throws LoginException if an app canceld the login process or the user is not enabled
+	 */
+	private function loginWithOAuth2($bearerToken) {
```
Ah - nice - you already did add this in here! Great!

DeepDiver1975 commented on lib/private/User/Session.php:
```diff
@@ -516,6 +520,59 @@ private function loginWithToken($token) {
+	 * @throws LoginException if an app canceld the login process or the user is not enabled
+	 */
+	private function loginWithOAuth2($bearerToken) {
+		$app = new App('oauth2');
```
This is what we don't want to have - core depending on an app is a no go.

We need some plugin mechanism to this ... something like

```php
namespace \OCP\Authentication;

/**
 * @returns null | IUser
 */
interface IAuthModule {
    public function auth(IRequest $request);
}
```

And classes implementing this interface can be loaded via info.xml - basically just like the TwoFactor plugins

DeepDiver1975 replied:
And within the UserSession instead if calling tryTokenLogin, tryBearerLogin, trySessionLogin ....

We just loop over all registered modules

joneug replied:
Ok. I will try that as soon as I have time.

joneug committed:
Remove reference to OAuth2 class in webdav.php

joneug commented:
> Unfortunately I did not have luck authenticating after that.

Now it works to authenticate using the `/remote.php/dav/` endpoint. So now the `dav` app does not depend on the `oauth2` app anymore.

It did not work at first because I did not add the `<authentication/>` type in `info.xml`.

joneug committed:
Add plugin mechanism for authentication

joneug committed:
Update PHPDoc in IAuthModule

joneug committed:
Update PHPDoc for auth function

joneug commented:
The dependencies between the core and the `oauth2` app are now fully eliminated.

> We need some plugin mechanism to this

I added `IAuthModule` as suggested.

> And classes implementing this interface can be loaded via info.xml - basically just like the TwoFactor plugins

I updated the `oauth2` app according to this. The `AuthModule` class is registered in `info.xml` under `auth-modules`.

> And within the UserSession instead if calling tryTokenLogin, tryBearerLogin, trySessionLogin ....
> We just loop over all registered modules

I added the function `tryAuthModuleLogin` that implements this loop. I don't see a way of replacing `tryTokenLogin` with this loop as tokens are not implemented with an app. What do you think?

joneug committed:
Merge branch 'master' of https://github.com/owncloud/core into dav-oauth

# Conflicts:
#	lib/private/legacy/api.php

joneug committed:
Add getUserPassword function in IAuthModule

joneug committed:
Add loading of additional AuthBackends for webdav interface

joneug commented:
I changed the title of the PR, because now OAuth 2.0 is fully capsulated in the app and this PR addresses the more general problem of adding 3rd-party authentication backends.

Furthermore I added the function `getUserPassword` to the `IAuthModule` interface in order to be able to trigger the hooks. Where to get the password from should not concern this PR.

As a side note: in the [OAuth 2.0 app](https://github.com/owncloud/oauth2) an empty string is returned because we are not handling any passwords at all. Consequently, only master key encryption is working, but this restriction exists when using Shibboleth, too.

joneug commented:
Could you please review the changes?

Additionally considering a backport to `stable9.1` would be nice since the changes are relatively small and there is high interest form our side in using OAuth 2.0 in ownCloud 9.1.

DeepDiver1975 commented on lib/private/User/Session.php:
```diff
@@ -676,6 +681,79 @@ public function tryTokenLogin(IRequest $request) {
+	 */
+	public function tryAuthModuleLogin(IRequest $request) {
+		/** @var IAppManager $appManager */
+		$appManager = OC::$server->query('AppManager');
```
please inject appManager as parameter in the ctor

joneug replied:
I tried to inject the AppManager, but when changing the registration of `UserSession` in `Server.php` (see [here](https://github.com/pssl16/core/commit/cb6e56b9bc535e91466790f30549bd4de4e76ab7)) I get the following error:

```
PHP Fatal error:  Maximum function nesting level of '256' reached
```

Inside the closure for the `registerService` function, `$appManager = $c->getAppManager();` is called over and over again. Any ideas?

joneug replied:
No ideas?

DeepDiver1975 replied:
okay then let's keep it this way

DeepDiver1975 commented on lib/private/User/Session.php:
```diff
@@ -676,6 +681,79 @@ public function tryTokenLogin(IRequest $request) {
+						}
+
+						/** @var IAuthModule $authModule */
+						$authModule = OC::$server->query($class);
```
adding a type check?

```php
if ($authModule instanceof IAuthModule) {

joneug replied:
Done

joneug committed:
Add type check for IAuthModule

DeepDiver1975 commented:
hmmm ... two tests are failing on oracle ....

````
23:15:18 There were 2 failures:
23:15:18 
23:15:18 1) Test\Files\External\Service\GlobalStoragesServiceTest::testUpdateStorage with data set #1 (array('mountpoint', 'identifier:\Test\Files\Extern...ackend', 'identifier:\Auth\Mechanism', array('value1', 'value2', 'testPassword'), array('user1', 'user2'), array(), 15))
23:15:18 Failed asserting that two arrays are equal.
23:15:18 --- Expected
23:15:18 +++ Actual
23:15:18 @@ @@
23:15:18  Array (
23:15:18 -    0 => 'user1'
23:15:18 -    1 => 'user2'
23:15:18 +    0 => 'user2'
23:15:18 +    1 => 'user1'
23:15:18  )
23:15:18 
23:15:18 /var/lib/jenkins/workspace/owncloud-core_core_PR-26742-JCGXUYPGPX2CBJQVJBQBPP3QLMW2OPW6JKKJVSVR77ZQ5M6LTFKQ/tests/lib/Files/External/Service/GlobalStoragesServiceTest.php:185
23:15:18 
23:15:18 2) Test\Files\External\Service\GlobalStoragesServiceTest::testUpdateStorage with data set #3 (array('mountpoint', 'identifier:\Test\Files\Extern...ackend', 'identifier:\Auth\Mechanism', array('value1', 'value2', 'testPassword'), array('user1', 'user2'), array('group1', 'group2'), 15))
23:15:18 Failed asserting that two arrays are equal.
23:15:18 --- Expected
23:15:18 +++ Actual
23:15:18 @@ @@
23:15:18  Array (
23:15:18 -    0 => 'user1'
23:15:18 -    1 => 'user2'
23:15:18 +    0 => 'user2'
23:15:18 +    1 => 'user1'
23:15:18  )
23:15:18

joneug committed:
Merge branch 'master' of https://github.com/owncloud/core into dav-oauth

joneug commented:
I pulled the changes in `master` branch and now the Jenkins build was successful. But the travis build failed:

```
$ sh -c "if [ '$TEST_DAV' = '1' ]; then bash apps/dav/tests/travis/$TC/install.sh; fi"
bash: apps/dav/tests/travis/selenium/install.sh: No such file or directory

The command "sh -c "if [ '$TEST_DAV' = '1' ]; then bash apps/dav/tests/travis/$TC/install.sh; fi"" failed and exited with 127 during .
```

Are there errors in the build script (see changes of commit [cac897b](https://github.com/owncloud/core/commit/cac897be7f1bbc4b63d684e70a4caa80ada8b608))?

DeepDiver1975 commented:
> I pulled the changes in master branch and now the Jenkins build was successful. But the travis build failed:

I'd say this is unrelated ..... let's merge

DeepDiver1975 commented:
Thanks a lot! @joneug and your team! Great job!

joneug commented:
What do you think about backporting to 9.1?

> Additionally, considering a backport to stable9.1 would be nice since the changes are relatively small and there is high interest form our side in using this mechanism for implementing OAuth 2.0 in ownCloud 9.1.

DeepDiver1975 commented:
> What do you think about backporting to 9.1?

@PVince81 what's your opionion? THX

joneug commented:
I created a backport in PR #27370.
